### ISSUE TITLE
Multi-label prediction: dev.tsv and test.tsv have -> MISMATCH!

### ISSUE THREAD

**Issue by ChessMateK** [flairNLP/flair#2630]
Hello everyone!

I'm trying to use flair-0.9 (not the the latest version because of #2624) for multi-label predictions, in detail I have two labels (Label1 and Label2), whose values can be 0 or 1. In particular, train/dev/test are in the FastText format: 

> __label__y1 __label__y2 "Some text"

where y1 and y2 can be 0 or 1 (combinations are (0,0), (0,1), (1,0), (1,1), perfectly balanced).

Hence I set: 

> classifier = TextClassifier(document_embeddings, label_dictionary=my_label_dictionary, multi_label=True, label_type='Label1', loss_weights=weight_dict)

where `my_label_dictionary = [b'0', b'1']` but I do not understand if it is correct (because Flair gives only a list of all the possible labels) or not (because Flair should give 0 and 1 twice: the first time for the first label and the second time for the second label). 
In addition it is not completely clear how I should set `label_type` in this case.

When i look into dev.tsv and test.tsv I find for instance:

1. One label prediction seems missing:
```
- Gold: [1 (1.0), 1 (1.0)] 
- Pred: [1 (0.9993)] 
-> MISMATCH!
```

2. One label is wrongly predicted:
```
 - Gold: [0 (1.0), 0 (1.0)]
 - Pred: [0 (0.9952), 1 (0.6607)]
 -> MISMATCH!
```

3. Both labels are correctly predicted without "MISMATCH!":
```
- Gold: [0 (1.0), 1 (1.0)] 
- Pred: [0 (0.9937), 1 (0.6664)]
```

4. Both labels are wrongly predicted without "MISMATCH!":
```
 - Gold: [1 (1.0), 0 (1.0)]
 - Pred: [0 (0.937), 1 (0.9926)]
```

5. I never found "MISMATCH!" associated with the case of both labels wrongly predicted.

To wrap up, I am pretty sure I am doing something wrong, missing some settings out there. So my questions are:

- Can anyone help me with a complete overview of all the needed settings to do multi label prediction starting from the (CSV)ClassificationCorpus call also specifying which variables are expected to change respect to single label scenario (to look at through whatcher during debug)?
- I need also another clarification: is there any setting that allow me to set the learning of the neural network to minimize the loss with regard to all the labels (that is the standard multi-label scenario I'm expect Flair implements) or specifying a weight for each one (i.e. the prediction of some labels is more important than others)?

Thank you in advance.

**Comment by ChessMateK** [comment]
Hi everyone. I've an update:

1) The most important thing I noticed, that help me to "reduce" the problem, is related to the FastText format itself: it is not important in which column the label is located, but which is its name. What do I mean? Let's see my previous structure:

> __label__0 __label__0 "Some text 00"
> __label__0 __label__1 "Some text 01"
> __label__1 __label__0 "Some text 10"
> __label__1 __label__1 "Some text 11"

These are all the possible cases but there is a problem: the possible labels to be predicted result to be "0" or "1", not related to the columns. This means that the classifier predicts 0 or 1, not 0 or 1 for label1 and 0 or 1 for label2.

To solve this problem I changed the labels this way:

> __label__firstlabel-value0 __label__secondlabel-value0 "Some text 00"
> __label__firstlabel-value0 __label__secondlabel-value1 "Some text 01"
> __label__firstlabel-value1 __label__secondlabel-value0 "Some text 10"
> __label__firstlabel-value1 __label__secondlabel-value1 "Some text 11"

where "firstlabel-value0" and "firstlabel-value1" are the 0s and 1s related to the first label and "secondlabel-value0" and "secondlabel-value1" are the 0s and 1s related to the second label.

How did the results change?

1. This is a case that seems to me "possible" because it signals "MISMATCH" when the prediction is different from the gold.
```
 - Gold: [firstlabel-value0 (1.0), secondlabel-value1 (1.0)]
 - Pred: [firstlabel-value0 (1.0), secondlabel-value0 (0.5607)]
 -> MISMATCH!
```

2. These cases, instead, gave me the idea that the flair classifier was not working properly (or at least not as I expected). The rationale to display "MISMATCH" is linked to every prediction that is different from the gold, and this is ok, but the working method allows to predict values not related to the label column: only one label or both but in an inverted way.
```
 - Gold: [firstlabel-value1 (1.0), secondlabel-value1 (1.0)]
 - Pred: [secondlabel-value1 (0.9996)]
 -> MISMATCH!
```

```
 - Gold: [firstlabel-value1 (1.0), secondlabel-value1 (1.0)]
 - Pred: [secondlabel-value1 (0.9999), firstlabel-value0 (0.8346)]
 -> MISMATCH!
```

Overall, this behaviour seems to be linked to the way the FastText format is implemented.

I hope this helps.

Thank you.

**Comment by alanakbik** [comment]
Hello @ChessMateK thanks for sharing these observations and excuse the late reply. Yes you're right: The classifier is essentially a multi-label classifier, so it makes all label predictions completely independently. If there's enough training data and it learns well, it should still work reasonably well (at least that's what we saw in other problems in which we had a label hierarchy that we just modeled as a multi-label problem). 

We don't currently have direct support for hierarchical classification in Flair. You could train three classifiers (one for the first 1-0 classification, two more for the subclasses) and manually arrange them in a pipeline, but that would be a good amount of engineering overhead.
